pkg/controller/drain.go

```diff
@@ -487,87 +493,110 @@ func filterSharedPVs(pvMap map[string][]string) {
+		remainingPods []*api.Pod
+		fastTrack     bool
+		nretries      = int(o.MaxEvictRetries)
+	)
+
+	if attemptEvict {
+		for i := 0; i < nretries; i++ {
+			remainingPods, fastTrack = o.evictPodsWithPVInternal(attemptEvict, pods, volMap, policyGroupVersion, getPodFn, returnCh)
+			if fastTrack || len(remainingPods) == 0 {
+				//Either all pods got evicted or we need to fast track the return (node deletion detected)
+				break
 			}
 
-			glog.V(3).Info("waiting for PVs to detach from node for pod: ", pod.Name)
-			// Eviction was successful. Wait for pvs for this pod to detach
-			pvs := volMap[pod.Namespace+"/"+pod.Name]
-			err = o.waitForDetach(pvs, o.nodeName, stopCh)
-
-			if apierrors.IsNotFound(err) {
-				glog.V(3).Info("Node not found anymore")
-				returnCh <- nil
-				remainingPods = len(pods) - (i + 1) + len(retryPods)
-				// The node itself is gone. Stop evicting pods.
-				// We should anyway return successful eviction for the remaining pods
-				break EvictingPods
-			} else if err != nil {
-				glog.Errorf("Error when waiting for volume to detach from node. Err: %v", err)
-				returnCh <- err
-				continue
-			}
-			glog.V(3).Info("Detached pv for pod: ", pod.Name)
-			returnCh <- nil
+			glog.V(4).Info("Eviction for some pods will be retried after ", PodEvictionRetryInterval)
+			pods = remainingPods
+			time.Sleep(PodEvictionRetryInterval)
 		}
 
-		if len(retryPods) == 0 {
-			// There are no pods to retry
-			break
+		if !fastTrack && len(remainingPods) > 0 {
+			// Force delete the pods remaining after evict retries.
+			pods = remainingPods
+			remainingPods, _ = o.evictPodsWithPVInternal(false, pods, volMap, policyGroupVersion, getPodFn, returnCh)
 		}
-
-		glog.V(4).Info("Eviction for some pods will be retried after ", PodEvictionRetryInterval)
-		pods = retryPods
-		time.Sleep(PodEvictionRetryInterval)
+	} else {
+		remainingPods, _ = o.evictPodsWithPVInternal(false, pods, volMap, policyGroupVersion, getPodFn, returnCh)
 	}
 
-	for j := 0; j < remainingPods; j++ {
+	// Placate the caller by returning the nil status for the remaining pods.
+	for _, pod := range remainingPods {
 		glog.V(4).Info("Returning success for remaining pods")
 		// This is executed when node is not found anymore.
 		// Return success to caller for all non-processed pods so that the caller function can move on.
-		returnCh <- nil
+		if fastTrack {
+			returnCh <- nil
+		} else if attemptEvict {
+			returnCh <- fmt.Errorf("Error evicting pod %s/%s", pod.Namespace, pod.Name)
+		} else {
+			returnCh <- fmt.Errorf("Error deleting pod %s/%s", pod.Namespace, pod.Name)
+		}
 	}
 
 	return
 }
 
+func (o *DrainOptions) evictPodsWithPVInternal(attemptEvict bool, pods []*corev1.Pod, volMap map[string][]string,
+	policyGroupVersion string,
+	getPodFn func(namespace, name string) (*api.Pod, error),
+	returnCh chan error) (remainingPods []*api.Pod, fastTrack bool) {
+	var retryPods []*api.Pod
+	for i, pod := range pods {
+		var err error
+		if attemptEvict {
```

Comment:
```suggestion
		if attemptEvict && len(policyGroupVersion) > 0 {
```

Reply:
With regards to handling k8s versions that don't support eviction?

Reply:
Already covered [above](https://github.com/gardener/machine-controller-manager/pull/288/files#diff-face99aebd3cd79577003f7df747c863R341).